server/app/api/v1/charity/raffle/controller.js

```diff
@@ -0,0 +1,218 @@
+          purchase => purchase.buyer === user._id
+        )[0];
+        return res.json({
+          count: buyer.count,
+          user: {
+            email: user.email,
+            firstName: user.firstName,
+            lastName: user.lastName
+          }
+        });
+      })
+      .catch(err => next(err));
+  } else {
+    next(new NotPermittedError(
+      'Tickets cannot be purchased after raffle has ended.'
+    ));
+  }
+
+  function findByUserId() {
+    const userId = req.query.userId;
+
+    if (!userId.match(VALID_OBJECT_ID)) {
+      return next(
+        new ValidationError(
+          'charityId must be valid ObjectId',
+          VALID_OBJECT_ID,
+          'charityId'
+        )
+      );
+    }
+    return User.findById(userId)
+      .then((user) => {
+        if (!user) {
+          throw new NotFoundError('User not found');
+        } else {
+          return user;
+        }
+      });
+  }
+
+  function createUser() {
+    return new User({
+      firstName: req.body.firstName,
+      lastName: req.body.lastName,
+      email: req.body.email
+    }).save();
+  }
+
+  function addTickets(user) {
+    // checks for if user already purchased raffles
+    const purchaseIndex = findIndex(req.raffle.purchases, ['buyer', user._id]);
+    const userRafflesIndex = findIndex(user.raffles, req.raffle._id);
+
+    if (purchaseIndex > -1) {
+      // remove duplicate purchase from purchases array
+      // this is a workaround to an Mongoose error I was receiving
+      // See: https://github.com/Automattic/mongoose/issues/3816
+      const dupPurchase = req.raffle.purchases.splice(purchaseIndex, 1)[0];
+      req.raffle.purchases.push({
+        count: dupPurchase.count + Number(req.body.count),
+        buyer: user._id
+      });
+    } else {
+      req.raffle.purchases.push({
+        count: Number(req.body.count),
+        buyer: user._id
+      });
+    }
+    if (userRafflesIndex > -1) {
+      return [req.raffle.save(), user];
+    } else {
+      user.raffles.push(req.raffle);
+      return [req.raffle.save(), user.save()];
+    }
+  }
+};
+
+/**
+ * Draw winner
+ */
+const drawWinner = (req, res, next) => {
+  if (!req.raffle.winner && req.raffle.isActive) {
+    const winner = req.raffle.drawWinningTicket();
+    return merge(req.raffle, { winner })
+      .save()
+      .then(
+        (savedRaffle) => (
+          Raffle.populate(savedRaffle, {
+            path: 'winner'
+          })
+        )
+      )
+      .then(({ winner }) => res.json(winner))
+      .catch(err => next(err));
+  } else {
+    next(new NotPermittedError(
+      'Winner has already been chosen or raffle has ended.'
+    ));
+  }
+};
+
+/**
+ * Update Raffle
+ * To end raffle use isActive=false
+ */
+const update = (req, res, next) => {
+  return merge(req.raffle, req.body).save()
+    .then(savedRaffle => res.json(savedRaffle))
+    .catch(err => next(err));
+};
+
+export default {
+  params, all, get, create, purchase, drawWinner, update
+};
```

Comment:
Yeah there's not any ticket limits or checks going on in here or in the model. We can always add that constraint later on, if need be.